pkg/pub_dartdoc/lib/src/pub_hooks.dart

```diff
@@ -0,0 +1,195 @@
+// Copyright (c) 2021, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:convert';
+import 'dart:io' as io;
+
+import 'package:analyzer/dart/element/element.dart';
+import 'package:analyzer/file_system/file_system.dart';
+import 'package:analyzer/file_system/memory_file_system.dart';
+// ignore: implementation_imports
+import 'package:analyzer/src/generated/sdk.dart';
+// ignore: implementation_imports
+import 'package:analyzer/src/generated/source.dart';
+import 'package:dartdoc/dartdoc.dart';
+import 'package:path/path.dart' as p;
+import 'package:watcher/watcher.dart';
+
+const _maxFileCount = 10000;
+const _maxTotalLengthBytes = 10 * 1024 * 1024;
+
+/// Creates an overlay file system with binary file support on top
+/// of the input sources.
+class PubResourceProvider implements ResourceProvider {
```

Comment:
Why not just create a proper overlay, instead of relying on output folder detection?

When writing we always write to `_memoryResourceProvider`, when reading we read from `_memoryResourceProvider` and if not present there, then we check: `_defaultProvider`.

I guess the complicated part is when listing folders, we'll need to wrap and merge them... But it would be less magic..

We can also leave that as a TODO... (I'm guessing we can almost just steal the overlay implementation from analyzer and binary file support to it, why not?

Reply:
++